src/zkevm_specs/tx.py

```diff
@@ -0,0 +1,364 @@
+    # The columns are: (is_enabled, input_rlc, input_len, output_rlc)
+    table: Set[Tuple[FQ, FQ, FQ, FQ]]
+
+    def __init__(self):
+        self.table = set()
+        self.table.add((FQ(0), FQ(0), FQ(0), FQ(0)))  # Add all 0s row
+
+    def add(self, input: bytes, randomness: FQ):
+        output = keccak(input)
+        self.table.add(
+            (FQ(1), RLC(input, randomness).value, FQ(len(input)), RLC(output, randomness).value)
+        )
+
+    def lookup(self, is_enabled: FQ, input_rlc: FQ, input_len: FQ, output_rlc: FQ, assert_msg: str):
+        assert (is_enabled, input_rlc, input_len, output_rlc) in self.table, (
+            f"{assert_msg}: {(is_enabled, input_rlc, input_len, output_rlc)} "
+            + "not found in the lookup table"
+        )
+
+
+class WrongFieldInteger:
+    """
+    Wrong Field arithmetic Integer, representing the implementation at
+    https://github.com/appliedzkp/halo2wrong/blob/master/integer/src/integer.rs
+    """
+
+    limbs: bytes  # Little-Endian bytes
+
+    def __init__(self, value: int) -> None:
+        self.limbs = value.to_bytes(32, "little")
+
+
+class Secp256k1BaseField(WrongFieldInteger):
+    """
+    Secp256k1 Base Field.
+    """
+
+    def __init__(self, value: int) -> None:
+        WrongFieldInteger.__init__(self, value)
+
+
+class Secp256k1ScalarField(WrongFieldInteger):
+    """
+    Secp256k1 Scalar Field.
+    """
+
+    def __init__(self, value: int) -> None:
+        WrongFieldInteger.__init__(self, value)
+
+
+class ECDSAVerifyChip:
+    """
+    ECDSA Signature Verification Chip.  This represents an ECDSA signature
+    verification Chip as implemented in
+    https://github.com/appliedzkp/halo2wrong/blob/master/ecdsa/src/ecdsa.rs
+    """
+
+    signature: Tuple[Secp256k1ScalarField, Secp256k1ScalarField]
+    pub_key: Tuple[Secp256k1BaseField, Secp256k1BaseField]
+    msg_hash: Secp256k1ScalarField
+
+    def __init__(
+        self,
+        signature: Tuple[Secp256k1ScalarField, Secp256k1ScalarField],
+        pub_key: Tuple[Secp256k1BaseField, Secp256k1BaseField],
+        msg_hash: Secp256k1ScalarField,
+    ) -> None:
+        self.signature = signature
+        self.pub_key = pub_key
+        self.msg_hash = msg_hash
+
+    @classmethod
+    def assign(cls, signature: KeyAPI.Signature, pub_key: KeyAPI.PublicKey, msg_hash: bytes):
+        self_signature = (Secp256k1ScalarField(signature.r), Secp256k1ScalarField(signature.s))
+        pub_key_bytes = pub_key.to_bytes()
+        pub_key_bytes_x, pub_key_bytes_y = pub_key_bytes[:32], pub_key_bytes[32:]
+        pub_key_x = int.from_bytes(pub_key_bytes_x, "big")
+        pub_key_y = int.from_bytes(pub_key_bytes_y, "big")
+        self_pub_key = (Secp256k1BaseField(pub_key_x), Secp256k1BaseField(pub_key_y))
+        self_msg_hash = Secp256k1ScalarField(int.from_bytes(msg_hash, "big"))
+        return cls(self_signature, self_pub_key, self_msg_hash)
+
+    def verify(self, is_enabled: FQ, assert_msg: str):
```

Comment:
NOTE: The current implementation of the ECDSA chip won't pass the constraints if the signature is invalid.  For this circuit we need to allow disabling an ECDSA chip based on witness data (for padding transactions).  So to implement this maybe we need to change the ECDSA chip so that it contains a cell that for example is `0` when the verification fails and `!= 0` when it succeeds, or something similar, so that we can build a constraint that passes on invalid verification (for padding).